### PR TITLE
Remember and draw previously seen tiles

### PR DESCRIPTION
**Pull request by dvolk**
If the player has seen a tile but isn't seeing it currently this draws it as unlit terrain, rather than empty space.

https://i.imgur.com/quZnlYZ.png

### CODE REVIEW

File: src/mapdata.h [hunk 1]
```diff
@@ -447,6 +447,8 @@ struct submap {
     trap_id         trp[SEEX][SEEY];  // Trap on each square
     int             rad[SEEX][SEEY];  // Irradiation of each square
 
+    bool            known[SEEX][SEEY]; // has the player ever seen this tile
```
Comment (BevapDin) [inline src/mapdata.h]: Ui, this means telepathic knowledge transfer between all the players on the same world. What mutation allows that?

File: src/map.cpp [hunk 2]
```diff
@@ -1327,6 +1327,28 @@ ter_id map::ter(const int x, const int y) const
...
+    }
+
+    int lx, ly;
+    submap * const current_submap = get_submap_at(p, lx, ly);
+    return current_submap->known[lx][ly];
+}
+
+void map::set_pl_known( const tripoint &p )
+{
+    if( !INBOUNDS(p.x, p.y) ) {
```
Comment (BevapDin) [inline src/map.cpp]: What about `p.z`? There is now a `inbounds` function that accepts a single tripoint, use that instead.

File: src/map.cpp [hunk 3]
```diff
@@ -4727,10 +4755,18 @@ void map::draw(WINDOW* w, const point center)
...
            (dist > sight_range && LL_LOW == lit),
            LL_BRIGHT == lit);
+
+    // mark as seen
+    set_pl_known(p);
+
+   } else if (pl_knows(p)) {
+       // draw tiles that the player has seen previously but doesn't see right now
+       // as unlit and hide the items
+       drawsq(w, g->u, p, false, false, center.x, center.y, true, false);
```
Comment (BevapDin) [inline src/map.cpp]: Btw. if somebody or something changes those unseen tiles, the new content will be shown to the player - magic!

File: src/map.cpp (hunk 2 above)
Comment (dvolk) [inline src/map.cpp]: I missed it, thanks.

File: src/mapdata.h (hunk 1 above)
Comment (dvolk) [inline src/mapdata.h]: I'm not sure what you mean. There's only 1 player and no way to switch to a different creature as far as I know.

Comment (BevapDin) [inline src/mapdata.h]: You can save the game and start a new one with a different character in the very same world.

There is also the possibility of map sharing, see http://smf.cataclysmdda.com/index.php?topic=9156, which usually means several human players use the same map data.

Comment (dvolk) [inline src/mapdata.h]: That's not a problem since the array isn't saved.

Comment (BevapDin) [inline src/mapdata.h]: I stand corrected, that avoids the multiplayer problem completely.

Losing that knowledge through save & reload might be unexpected for the player, but it's probably acceptable.

File: src/map.cpp (hunk 3 above)
Comment (dvolk) [inline src/map.cpp]: Thanks for catching that. I fixed it by caching the symbol.

**Comment by kevingranade** [comment]
At a high level, I think this is one of the rare things that does need an option to toggle it, some people are just going to prefer one or the other.
Implementation-wise, I don't like the known data being stored on the submap, regardless of whether it's persistent, it needs to live with the player object.  If it's non-persistent a simple reality-bubble sized array of characters would be sufficient (though you'd have to shift it on map shift).  A more general solution would be something like a std::unordered_map<tripoint, known_data> with the points being in absolute coordinates.  Then you never have to worry about shifting the contents as the player moves around.  This would also get us pretty close to a similar feature where you can recall the location of every item you've ever seen

**Comment by Rivet-the-Zombie** [comment]
No matter how it's accomplished, I'd like to see this happen.

**Comment by dvolk** [comment]
I added it to options.

The player obj would be a more logical place to put it, but it has the downside that it would involve duplicating a lot of code. I'm not sure it's worth it.

**Comment by Rivet-the-Zombie** [comment]
I like this one a lot.

**Comment by kevingranade** [comment]
How much code would you be duplicating?  As I outlined, you need code to do the relative to absolute translation (which is a pain to sort out, but a solved problem) and a container in the player object.  AFAIK that's it.

**Comment by dvolk** [comment]
At a minimum it would also need to be dynamically created and destroyed as you move around, which would mean coupling it to the submap with callbacks or rewriting that part of the code. I don't see the advantage of it over just keeping the data in the submap.

**Comment by BevapDin** [comment]
> At a minimum it would also need to be dynamically created and destroyed as you move around, which would mean coupling it to the submap with callbacks or rewriting that part of the code.

You can keep it around for as long as the game runs (and not saving it). A loaded game will not have the data. Or you could destroy it when the map shifts (`game::vertical_move` or `game::update_map`). Through, I like the submap implementation (-:

**Comment by tivec** [comment]
> No matter how it's accomplished, I'd like to see this happen.

I second that emotion. It has to be persistent over game/map loading though, or it doesn't make much sense at all. I also agree that it should be an option. 

Just a thought, even though it is dimmed out, it would be nice if it was even darker grey, if at all possible.

**Comment by sparr** [comment]
I'm a fan of this, nethack style.

**Comment by kevingranade** [comment]
Seconding the request for distinct coloring, but I don't think there's a
color available to use in the terminal.

**Comment by sparr** [comment]
I'd like an option to have it just all appear in light or dark grey. I know it's ambiguous, but the color space is already overloaded, and the remaining options (like giving remembered tiles a blue background) would look terrible imo.

Perhaps a hotkey to toggle remembered tiles between correct color and grey? When it's accurate color, you get the most info, but you can't tell the edge of your current view bubble. When it's grey, you get less info, but you can tell where your view bubble ends. (this would have to interact with the current "low light" system in some way)

**Comment by dvolk** [comment]
I think it's possible to define more colors, but support for it would depend on the terminal emulator.

**Comment by i2amroy** [comment]
>I think it's possible to define more colors, but support for it would depend on the terminal emulator.

Which windows is notoriously bad about sadly. :/

**Pull request closed without merging**